crates/containerd-shim-wasm/src/sandbox/libcontainer_instance.rs

```diff
@@ -0,0 +1,187 @@
+//! Abstractions for running/managing a wasm/wasi instance that uses youki's libcontainer library.
+
+use anyhow::Context;
+use chrono::Utc;
+use libc::{SIGINT, SIGKILL};
+use libcontainer::{
+    container::{Container, ContainerStatus},
+    signal::Signal,
+};
+use log::error;
+use nix::{
+    errno::Errno,
+    sys::wait::{waitid, Id as WaitID, WaitPidFlag, WaitStatus},
+};
+
+use crate::sandbox::{
+    instance::ExitCode,
+    instance_utils::{get_instance_root, instance_exists},
+};
+
+use crate::sandbox::InstanceConfig;
+use std::{path::PathBuf, thread};
+
+use super::{error::Error, instance::Wait, Instance};
+
+/// LibcontainerInstance is a trait that gets implemented by a WASI runtime that
+/// uses youki's libcontainer library as the container runtime.
+/// It provides default implementations for some of the Instance trait methods.
+/// The implementor of this trait is expected to implement the
+/// * `new_libcontainer()`
+/// * `get_exit_code()`
+/// * `get_id()`
+/// * `get_root_dir()`
+/// * `build_container()`
+/// methods.
+pub trait LibcontainerInstance {
+    /// The WASI engine type
+    type E: Send + Sync + Clone;
+
+    /// Create a new instance
+    fn new_libcontainer(id: String, cfg: Option<&InstanceConfig<Self::E>>) -> Self;
+
+    /// Get the exit code of the instance
+    fn get_exit_code(&self) -> ExitCode;
+
+    /// Get the ID of the instance
+    fn get_id(&self) -> String;
+
+    /// Get the root directory of the instance
+    fn get_root_dir(&self) -> Result<PathBuf, Error>;
+
+    /// Build the container
+    fn build_container(&self) -> Result<Container, Error>;
```

Comment:
`build_container` seems like a fairly heavy trait function.

I think if we add something like `open_stdio` to the trait we can implement common `build_container`?

Reply:
IIUC, that would move the stdio redirection work from the executor to the instance.
Is there any instance / executor that would not want to do the stdio redirection trick (slight/spin/wws)?
Can that be handled by default in `Instance` instead of `LibcontainerInstance`?